wrappers/desktop.go

```diff
@@ -190,6 +190,47 @@ func updateDesktopDatabase(desktopFiles []string) error {
 	return nil
 }
 
+// desktopPrefix returns the prefix string for the desktop files that
+// belongs to the given snapInstance. We need to do something custom
+// here because a) we need to be compatible with the world before we had
+// parallel installs b) we can't just use the usual "_" parallel installs
+// separator because that is already used as the separator between snap
+// and desktop filename.
+func desktopPrefix(s *snap.Info) string {
+	if s.SnapName() == s.InstanceName() {
+		return s.SnapName()
+	}
+	// we cannot use the usual "_" separator because that is also used
+	// to separate "$snap_$desktopfile"
+	return fmt.Sprintf("%s+%s", s.SnapName(), s.InstanceKey)
+}
+
+func installedDesktopFileName(s *snap.Info, rawcontent []byte, df string) (string, error) {
+	var snapApp string
+	var err error
+
+	scanner := bufio.NewScanner(bytes.NewReader(rawcontent))
+	for i := 0; scanner.Scan(); i++ {
+		bline := scanner.Bytes()
+		if bytes.HasPrefix(bline, []byte("Exec=")) {
+			snapApp, _, err = rewriteExecLine(s, df, string(bline))
+			if err != nil {
+				return "", err
+			}
+		}
+	}
+
+	for i := 0; i < 1000; i++ {
+		dfBaseName := fmt.Sprintf("%s_%s_%v.desktop", desktopPrefix(s), snapApp, i)
```

Comment:
I think this may end up a little confusing in the end. For instance when manipulating mime handlers via `xdg-mime` you need to use desktop file names. If you have 2 desktop files, say `foo.desktop` and `foo-some-handler.desktop`, both for app `foo` they would end up being named `snap_foo_0.desktop` and `snap_foo_1.desktop`. IIRC `filepath.Glob` (as used by `AddSnapDesktopFiles`) result order is not deterministic, so there's a chance you'd end up with different names for each install.